components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt

```diff
@@ -311,24 +391,13 @@ data class SyncableLoginsStore(
      * @throws [InvalidKeyException] if the provided [key] isn't valid.
      */
     suspend fun <T> withUnlocked(block: suspend (AsyncLoginsStorage) -> T): T {
-        // Instead of synchronizing on 'store' let's be optimistic and just swallow mismatched locks.
 
-        if (store.isLocked()) {
-            try {
-                store.unlock(key().await()).await()
-            } catch (e: MismatchedLockException) {
-                // Oh well, it's already unlocked!
-            }
-        }
+        store.ensureUnlocked(key().await()).await()
```

Comment:
At face value, this looks unsafe to me. Does anything prevent another thread from locking the store after you've ensured that it's unlocked and before you operate on it?

I'd prefer to see something like...
```
store.whileUnlocked(key().await()) {
    return block(store)
}
```
... and let the underlying implementation deal with locking, mutexing, whatever.

Reply:
After discussion in slack, this is tricky and should be done in a follow up if desired.